### ISSUE TITLE
controller does not accept a Service with Type: "LoadBalancer"

### ISSUE THREAD

**Issue by iljaweis** [F5Networks/k8s-bigip-ctlr#478]
#### Description
k8s-bigip-ctlr requires that the Kubernetes "Services" it creates virtual servers for are of type "NodePort". If I change the type to "LoadBalancer", the corresponding virtual server and its pool and pool members are not created and are even deleted if they existed before and I change type from NodePort to LoadBalancer.

Is there a reason that type "LoadBalancer" can / should not be used for the services?

I'm asking because I am currently writing a controller that integrates k8s-bigip-ctlr with an IP address management for virtual server addresses. End user creates a Service, my controller assigns a new address and creates/annotates the k8s-bigip-ctlr's ConfigMap. Having to use "NodePort" makes this method a bit clumsy because the Service then has to be annotated (with custom annotations), for example to signal that the virtual service should be managed by k8s-bigip-ctlr, and to later look up with virtual server address was assigned.

The type "LoadBalancer" supports all these using standard attributes and would be a cleaner and (I believe) more compatible way to deploy services that require external loadbalancing.

#### Kubernetes Version
v1.8.0

#### Controller Version
1.3.0

#### BIG-IP Version
BIG-IP 13.0.0 Build 0.0.1645 Final

**Comment by rsalsa** [comment]
Hi @iljaweis - The type [LoadBalancer]( https://kubernetes.io/docs/concepts/services-networking/service/#type-loadbalancer) for K8S services were designed for configuring load balancers in a cloud provider deployment (e.g. GKE, AKS, ELB, etc.) - K8S will call the underlying provider's API to configure the LB endpoint. Hence why we have to use NodePort for services. 

We support in the latest version K8S [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) resources which are a much more flexible and robust option for exposing K8S services externally. The supported annotations for our controller can be found in our [documentation](http://clouddocs.f5.com/products/connectors/k8s-bigip-ctlr/v1.3/#supported-annotations). 

HTH

**Comment by iljaweis** [comment]
HI @rsalsa,

thank you for your answer. I'm not quite sure how to apply it to my situation... ;-)

I agree about using an Ingress where possible. I (believe I) understand what type LoadBalancer is used for. Basically I'm trying to set up my installation (on premise, using BigIP) so that, from the user's perspective, it behaves exactly like in cloud provider environments when deploying a Service with type LoadBalancer:

- allocate an IP address
- create the loadbalancer / virtual service

I don't expect k8s-bigip-ctlr to do all that; I'm scanning for deployed Services and then create a ConfigMap using the http://clouddocs.f5.com/containers/v2/kubernetes/kctlr-manage-bigip-objects.html#assign-ip-addresses-to-big-ip-virtual-servers-using-ipam feature. k8s-bigip-ctlr behaves exactly like I expect and creates the correct virtual server for me.

What I would like to achieve is that my users do not have to bother with how to get an IP address for the virtual service, understand how exactly the BigIP integration works and can use the same Service across different infrastructures. I've already made progress writing a controller to do exactly that, but I noticed that I cannot use LoadBalancer because k8s-bigip-ctlr explicitly skips services that are not a NodePort. Right now I'm using NodePort with a number of custom annotations.

I believe LoadBalancer _should_ work for that because it behaves exactly like NodePort, especially creating a fixed port on each node.

Does this make any sense? Am I missing something, or is what's I'm trying to do generally a bad idea?
